iceoryx_hoofs/source/cli/command_line_argument_parser.cpp

```diff
@@ -0,0 +1,410 @@
+        }
+
+        uint64_t optionNameStart = (m_argv[i][1] == '-') ? 2 : 1;
+        auto optionEntry = m_optionSet->getOption(OptionName_t(cxx::TruncateToCapacity, m_argv[i] + optionNameStart));
+
+        if (!optionEntry)
+        {
+            switch (actionWhenOptionUnknown)
+            {
+            case UnknownOption::TERMINATE:
+            {
+                std::cout << "Unknown option \"" << m_argv[i] << "\"" << std::endl;
+                printHelpAndExit();
+                return m_optionValue;
+            }
+            case UnknownOption::IGNORE:
+            {
+                if (isNextArgumentAValue(i))
+                {
+                    skipCommandLineArgument();
+                }
+                continue;
+                break;
+            }
+            }
+        }
+
+        if (isOptionSet(*optionEntry))
+        {
+            return m_optionValue;
+        }
+
+        if (optionEntry->details.type == OptionType::SWITCH)
+        {
+            m_optionValue.m_arguments.emplace_back(*optionEntry);
+            m_optionValue.m_arguments.back().value.clear();
+        }
+        else
+        {
+            if (!doesOptionHasSucceedingValue(*optionEntry, i))
+            {
+                return m_optionValue;
+            }
+
+            if (!doesOptionValueFitIntoString(m_argv[i + 1]))
+            {
+                return m_optionValue;
+            }
+
+            m_optionValue.m_arguments.emplace_back(*optionEntry);
+            m_optionValue.m_arguments.back().value.unsafe_assign(m_argv[i + 1]);
+            skipCommandLineArgument();
+        }
+    }
+
+    setDefaultValuesToUnsetOptions();
+
+    if (m_optionValue.has("help") || !areAllRequiredValuesPresent())
+    {
+        printHelpAndExit();
+        return m_optionValue;
+    }
+
+    return m_optionValue;
+}
+
+bool CommandLineArgumentParser::doesOptionHasSucceedingValue(const OptionWithDetails& value,
+                                                             const uint64_t position) const noexcept
+{
+    bool doesOptionHasSucceedingValue = (position + 1 < m_argc);
+    if (!doesOptionHasSucceedingValue)
+    {
+        std::cout << "The option \"" << value << "\" must be followed by a value!" << std::endl;
+        printHelpAndExit();
+    }
+    return doesOptionHasSucceedingValue;
+}
+
+
+void CommandLineArgumentParser::setDefaultValuesToUnsetOptions() noexcept
+{
+    for (const auto& r : m_optionSet->m_availableOptions)
+    {
+        if (r.details.type != OptionType::OPTIONAL)
+        {
+            continue;
+        }
+
+        bool isOptionAlreadySet = false;
+        for (auto& option : m_optionValue.m_arguments)
+        {
+            if (option.isSameOption(r))
+            {
+                isOptionAlreadySet = true;
+                break;
+            }
+        }
+
+        if (!isOptionAlreadySet)
+        {
+            m_optionValue.m_arguments.emplace_back(r);
+        }
+    }
+}
```

Comment:
This also took me a while. Actually this does not set default values to unset options since all specified options are already in `m_optionValue.m_arguments`. What is does is adding unspecified options to the specified options if `actionWhenOptionUnknown` is not `UnknownOption::TERMINATE`. Please keep a consistent terminology. Currently it is a mixture of `unknown` and `unset`.

I would even argue to not add the unknown options to `m_optionValue.m_arguments`. If the developer forgot to specify the option at least it will be revealed in tests and the incorrect behavior can be fixed.